### ISSUE TITLE
Interactively rendering ~10 million points using datashader

### ISSUE THREAD

**Issue by ahnsws** [napari/napari#6148]
## 🚀 Feature
Hello everyone, I'm not sure what work's been done on optimizing the Points layer, but I thought I'd try and give a go at using [datashader](https://datashader.org/) to render into an image that napari can handle.

## Motivation
We work with large segmented images that have millions of cells, and visualizing them somehow would be great.

## Pitch
I threw together a script as a proof of concept, with the gif below:
![Peek 2023-08-14 13-24](https://github.com/napari/napari/assets/17461258/628d77c7-92b8-4f39-b464-6dbc9fd5c257)

Here's the script. I'm not really familiar with vispy or napari Layers, so please let me know if you see something that doesn't fit the intended API.

Edit: I moved the script to [this gist](https://gist.github.com/ahnsws/f128a0af3a898b57d5d70df80e34cd8f)

A few notes:
- The above is of 1e7 points, but it gets a little laggy. 1e6 points is pretty close to 60 fps.
- I have also tried rasterizing polygons with datashader, but it is significantly slower.
- datashader's shade and dynspread can improve the interpretation of dense parts of the image, but I didn't include them here because they ended up being too slow.
- Although datashader uses numba-optimized python code, maybe some better algorithms/dask parallelization on the spatialpandas/datashader side can improve performance.
- At high resolution, the individual points appear as single pixels; maybe it'd be possible to seamlessly switch between datashader's downsampled view and an alternative view with better rendering of points at certain zoom levels.

I'd love to hear your thoughts!

**Comment by melonora** [comment]
Hi @ahnsws,

I am sorry that so far we didn't reply yet. Looks awesome, thank you! Would you be willing to join one of the community meetings some time to discuss? You can find the schedule [here](https://napari.org/stable/community/meeting_schedule.html). Also, we have a [zulip](napari.zulipchat.com) where I also post the agenda just a bit before the meeting to which people can freely add items.

**Comment by psobolewskiPhD** [comment]
To echo what @melonora said: I'm one of the rocket emoji, but forgot to followup!
I hope you can make a community meeting to give us more context.
Looking at the datashader docs a smidge I almost see this as a different layer type than Points? More of a heat map? Could be really amazing to represent certain features, like classifier output, particularly on multiscale images?

**Comment by ahnsws** [comment]
Hi @melonora, I'd be happy to! Although the one tomorrow at 3am EST (my timezone) might be hard to join, so I will plan on joining the one next week (on 9/27).

Hi @psobolewskiPhD, yes exactly, this layer is kind of in between an Image and a Points... because there are so many more points than pixels when zoomed out, datashader acts as a 2D histogram, and I believe this is actually the intended use case. But I think it would be very useful if it acted as a true Points layer when zoomed in enough to see individual points.

I don't see why you couldn't use this to visualize other dynamically computed features based on the viewport dimensions and position; basically, anything that has the signature of the `_update_draw` method. Actually, I had hoped that millions of polygons (and not just points) could be visualized, which datashader does support, but is too slow for the time being, even with spatialpandas and its fancy spatial indexing. I wonder what qupath is doing there, since it is capable of buttery smooth interactivity even with millions of polygons.

Here are some docs that inspired me:

- [why we need something like datashader anyways](https://datashader.org/user_guide/Plotting_Pitfalls.html)
- [the nyc taxi dataset](https://examples.holoviz.org/nyc_taxi/nyc_taxi.html)

**Comment by melonora** [comment]
@ahnsws Awesome as I won't be able to join today either haha. I will add you to the agenda for next week.

**Comment by melonora** [comment]
I am familiar with the datashader in interactive plotting libraries like bokeh. I noticed there the same thing that it can get laggy when going above 10 million pixels. I could be wrong here but I believe this is because for creating the pixel buffer rendering it still depends on the whole dataset.
I see the point of it being inbetween Image and Points, though ultimately the data source is still points. 
@andy-sweet @kephale I think it would be interesting to see how this compares to tiled rendering / async. In any case the approach of datashader wouldn't require a chunked data format. What do you think?

**Comment by melonora** [comment]
Hi @ahnsws, just as a reminder, tomorrow at 8:30 am PT. Hope to see you there!

**Comment by ahnsws** [comment]
Hi @melonora, yes planning on it, thank you for the links!

**Comment by ahnsws** [comment]
@melonora thanks for inviting me to the community meeting. Here is some code for reproducing the polygons example, using the [NYC buildings dataset (zip; download warning)](http://s3.amazonaws.com/datashader-data/nyc_buildings.parq.zip).

```python
import napari
import numpy as np
import numpy.typing as npt
import xarray as xr
from datashader import Canvas
from napari._qt.layer_controls.qt_layer_controls_base import QtLayerControls
from napari._qt.layer_controls.qt_layer_controls_container import layer_to_controls
from napari._vispy.layers.base import VispyBaseLayer
from napari._vispy.utils.visual import layer_to_visual
from napari._vispy.visuals.image import Image as ImageNode
from napari.layers import Layer
from spatialpandas import GeoDataFrame
from spatialpandas.dask import DaskGeoDataFrame
from spatialpandas.io import read_parquet_dask
from vispy.visuals.transforms import ChainTransform, STTransform, MatrixTransform


class DatashaderGraphics(Layer):
    def __init__(self, gdf: GeoDataFrame):
        # fake data for now (not sure how to replace with a dataframelike obj)
        data = np.zeros(shape=(2, 2))

        super().__init__(data, ndim=2)

        self._gdf = gdf
        self._data = data
        self._view = np.zeros(shape=(1, 1))

    @property
    def data(self) -> npt.NDArray:
        return self._data

    # @data.setter
    # def data(self, data: npt.NDArray | None):
    #     self._data = data
    #     self._gdf = GeoDataFrame({"geometry": PointArray(data)})

    @property
    def _extent_data(self) -> npt.NDArray:
        x0, y0, x1, y1 = self._gdf.geometry.total_bounds
        return np.array([[y0, x0], [y1, x1]])

    @property
    def _data_view(self):
        return self._view

    def _update_draw(self, scale_factor, corner_pixels_displayed, shape_threshold):
        # this is where we can do anything to our heart's content
        width, height = shape_threshold

        miny, minx = corner_pixels_displayed[0]
        maxy, maxx = corner_pixels_displayed[1]

        x_range = (minx, maxx)
        y_range = (miny, maxy)

        canvas = Canvas(
            plot_width=width,
            plot_height=height,
            x_range=x_range,
            y_range=y_range,
        )
        # to change the way datashader datashades, pass a different agg method; currently "any"
        # img: xr.DataArray = canvas.polygons(self._gdf, geometry="geometry", agg=count())
        img: xr.DataArray = canvas.polygons(self._gdf, geometry="geometry")
        # img: xr.DataArray = canvas.points(self._gdf, geometry="geometry")

        #  we could include a few things here like datashader's `shade` and `dynspread`, but they are too slow...
        view = img.data
        # the "any" agg method results in a bool dtype
        if view.dtype == np.bool_:
            view = view.astype(np.uint8)

        # quick and dirty renormalization; replace with a sensible renormalization method
        max_view = view.max()
        if max_view == 0:
            self._view = view
        else:
            self._view = view / max_view

        # self._view = view
        self.events.set_data()

    def _get_ndim(self):
        return 2

    def _get_state(self):
        state = self._get_base_state()
        state.update({"ndim": self.ndim, "data": self.data})
        return state

    def _get_value(self, position) -> int | None:
        return None

    def _set_view_slice(self):
        pass

    def _update_thumbnail(self):
        self.thumbnail = np.zeros(self._thumbnail_shape)


class QtDatashaderGraphicsControls(QtLayerControls):
    def __init__(self, layer: DatashaderGraphics) -> None:
        super().__init__(layer)


class VispyCanvasLikeImageLayer(VispyBaseLayer):
    layer: DatashaderGraphics
    node: ImageNode

    def __init__(self, layer):
        super().__init__(layer, node=ImageNode())

    def _on_data_change(self):
        if not self.layer.loaded:
            # Do nothing if we are not yet loaded. Calling astype below could
            # be very expensive. Let's not do it until our data has been loaded.
            return

        self.node.set_data(self.layer._data_view)
        self.node.visible = self.layer.visible

        # do not translate or scale the scene
        transforms = ChainTransform(
            STTransform(scale=[1, 1, 1, 1], translate=[0, 0, 0, 0]),
            STTransform(scale=[1, 1, 1, 1], translate=[0, 0, 0, 0]),
            MatrixTransform(matrix=np.eye(4)),
        )
        self.node.transforms.scene_transform.transforms = transforms
        self.node.update()


def run():
    layer_to_controls[DatashaderGraphics] = QtDatashaderGraphicsControls
    layer_to_visual[DatashaderGraphics] = VispyCanvasLikeImageLayer

    gdf: DaskGeoDataFrame = read_parquet_dask("./path/to/nyc_buildings.parq")
    print(len(gdf))
    print(gdf.geometry.total_bounds)

    layer = DatashaderGraphics(gdf)

    viewer = napari.Viewer()
    viewer.add_layer(layer)
    napari.run()


if __name__ == "__main__":
    run()

```

**Comment by psobolewskiPhD** [comment]
Just out of curiosity, I tested 10M points in 0.5.6rc0.
My script:
```
import napari
import numpy as np
import time

NUM_POINTS=10_000_000

rng = np.random.default_rng(seed=0)
points = rng.standard_normal(size=(NUM_POINTS, 2)) * 1000

viewer = napari.Viewer()

start = time.time()
layer = napari.layers.Points(points)
create = time.time()
print("creating layer:", create - start)
viewer.add_layer(layer)
end = time.time()
print("Added layer:", end - create)

napari.run()
```
results on my 2020 M1:
```
creating layer: 1.658944845199585
Added layer: 6.519549131393433
```
Not bad at all!

Pan is a bit slow, zoom is pretty fine. Selection is fine. But moving a point blocks.
<img width="1198" alt="Image" src="https://github.com/user-attachments/assets/becc0c04-538f-4d29-9c15-adf4b7609e9f" />

(If you change to 1M points then everything is instant except moving a point, which is pretty laggy but doesn't block)

**Comment by ahnsws** [comment]
I've had some time to revisit this, especially now that we are attempting to do some spatial transcriptomics analysis with napari, and I was able to adapt the above code to visualize ~530M spots from a Nanostring CosMx run we did a while ago. It's **eyewateringly** slow (~10 seconds per frame, dropping to ~2 seconds when zoomed in), even with dask+parquet+spatialpandas, and despite my best efforts it still blocks the main thread, but it's an interesting way to visualize the data nonetheless, especially on top of registered image data. There's probably low-hanging fruit with regards to a more efficient method of spatial indexing somewhere. However, I'm honestly not sure of the utility of needing to interact with ~1e8 points this way...

![Image](https://github.com/user-attachments/assets/0ecd082d-b6f0-4f23-bf3a-b6b2c73f949b)

Zoomed in:

![Image](https://github.com/user-attachments/assets/77b3ad39-faad-4dd7-9b31-c1b657b4efb1)

I was inspired by [this post from Holoviz](https://examples.holoviz.org/gallery/ship_traffic/ship_traffic.html).

I think for ~1e7 points, the way @psobolewskiPhD had demonstrated above sounds perfect; perhaps having a "non-editable" mode to avoid inadvertently moving a point would be a nice-to-have.

**Comment by TimMonko** [comment]
Pinging @melonora as much of his focus right now is in this area, and I'd guess he has some thoughts. 
Also maaaybe @brisvag if I understand some of what he talked about today at the community meeting correctly. 

This looks like a tough challenge that I have little knowledge on, so sorry I can't say much more! Besides woohoo science!

**Comment by psobolewskiPhD** [comment]
@ahnsws Very awesome! I have end users I support at my day job that would be very interested in this.
When you say:
> I was able to adapt the above code to visualize ~530M spots from a Nanostring CosMx

which code do you mean?

**Comment by ahnsws** [comment]
@psobolewskiPhD ah I see, I was referring to [this gist](https://gist.github.com/ahnsws/f128a0af3a898b57d5d70df80e34cd8f), but it's a bit outdated now. I'm in the process of putting together a self-contained example using the Marine AIS dataset (although one issue I had was that the y-axis is positive down in napari, so all the routes are also flipped upside down).

**Comment by TimMonko** [comment]
> One issue I had was that the y-axis is positive down in napari, so all the routes are also upside down

Ah, something I can help with. There is wonderfully great news for you that makes this WAY easier. Thanks to @jni, napari 0.6.0 ships with camera orientation for _exactly_ this reason. Check out the alpha to play with it ([0.6.0a0](https://github.com/napari/napari/releases/tag/v0.6.0a0) just contains the API) 6.0.0a1 is coming soon. 

You are also welcome to play with all the orientation stuff on main. #7663 adds the API (which is in 6.0.0a0), #7686 a GUI to play with this in the viewer, #7770 the handedness property, and #7784 the orientation to preferences

For you, you'll want `viewer.camera.orientation = ('towards', 'up', 'right')` I'm pretty sure, though we've flipped things a bit since 0.5.6

**Comment by ahnsws** [comment]
Hi @TimMonko, thanks for the info! I just tried `viewer.camera.orientation = ('towards', 'up', 'right')` with `0.6.0a0` and unfortunately the scene goes purple (with viridis applied). I'll play around with it more and open a separate issue if needed; it's likely some bounds are just swapped. Not a big deal though since this specific dataset was just for testing. Thanks!

**Comment by TimMonko** [comment]
I think the colormap issue due to the calculation of x_range and y_range. It works with
```
x_range = (min(minx, maxx), max(minx, maxx))
y_range = (min(miny, maxy), max(miny, maxy))
```
but the data itself doesn't flip with the camera orientation. The data flips, but so does the entire canvas, I think. It's something that likely isn't hooked up like in other layers. A bit above me though ❤ I can look into it as well if you have any specific questions.

Partly just wanted to drop in that color fix so Juan didn't panic that orientation is broken. (or maybe it is, but like you said, probably the layer)

**Comment by ahnsws** [comment]
Yeah I agree, the way I've coded the boundaries in physical/world coordinates is a bit hacky. The geometries are already specified in physical/world coordinates, so I think the way I wrote how the `corner_pixels` are obtained needs to change with how napari makes them.

**Comment by brisvag** [comment]
> Pinging [@melonora](https://github.com/melonora) as much of his focus right now is in this area, and I'd guess he has some thoughts. Also maaaybe [@brisvag](https://github.com/brisvag) if I understand some of what he talked about today at the community meeting correctly.
> 
> This looks like a tough challenge that I have little knowledge on, so sorry I can't say much more! Besides woohoo science!

Mmm not sure this is, though I admit I didn't follow the whole conversation closely and I'm not familiar with datashader and why it's being used here :P For context, @TimMonko is referring to my planned work on instanced rendering (which should speed up rendering of a lot of similar objects), but this does not apply to points which are ffectively already using a similar system.

**Comment by TimMonko** [comment]
Thanks Lorenzo, I was more referring to 'trying to render lots of objects with applied transforms' which I know datashader isn't doing but maybe you had advice or ideas for trying to render effectively 530M objects on the canvas

**Comment by melonora** [comment]
Both instanced rendering and datashading have their different use cases. Here I would say that datashading is more suitable if the person viewing the data knows that datashading is being applied. A colormapping indicating how many gene targets are aggregated in a particular area is more informative then trying to show each particular gene target.

Maybe broader discussion would be should this be part of a plugin or should this be in core. I don't have time to play around with it now, but very cool! In case people think it should live in a plugin, I could incorporate it in `napari-spatialdata`.

@ahnsws if you would like to give `SpatialData` (https://spatialdata.scverse.org/en/stable/) a try for your data let me know! Would be happy to help out and would also appreciate any feedback

**Comment by ahnsws** [comment]
Hello all!

I did this initially out of curiosity, since when I first made this issue, it wasn't possible to plot 10M points in napari; now, as @psobolewskiPhD showed, this can now be done out of the box. However, now that we are regularly working with 100M and 1B points, and since it looked like datashader could manage this, I thought I'd give it a go.

Hi @melonora, I think this would work well as a plugin. I'd be more than happy to continue working on making this more efficient, as being able to render 100M-1B points this way is just so incredibly useful to us. If there are any suitable alternatives, please let me know (since I'd rather be working on the spatial biology itself haha...)

As for interpreting the datashaded image, it is effectively a distributed 2D histogram, and so I'd imagine it comes with the benefits and pitfalls of one. [This post from datashader](https://datashader.org/user_guide/Plotting_Pitfalls.html) goes into some of the downsides of plotting so many points that they overlap each other, hiding the distribution. As seen in @psobolewskiPhD's image with 10M points, it is unclear just how dense things are in the middle of the cluster, and without the code, I wouldn't have known that it was normally distributed. I agree with @melonora that for our particular use case, the distribution of points is indeed more informative than the individual points themselves.

Yes, I have heard of `spatialdata`! I haven't had a chance to use it yet but it looks very promising now that we have a few dozen CosMx datasets so far (with more on the way).

**Comment by melonora** [comment]
Great, let me know if you would like to set a meeting if you want to know about some of the developments we are planning or in general have questions!

**Comment by ahnsws** [comment]
I think that since the core of this issue has been resolved (for me, ~50M points using v0.5.6 is slightly sluggish but still absolutely works), I'll close this issue. I think if there is possible future work on dedicated out-of-memory scatter-plot rendering, that should deserve its own issue. Thank you everyone for the discussion!

**Comment by psobolewskiPhD** [comment]
I'm going to reopen because the datashader approach is cool and workable. Open issue is easier to find!

**Comment by jni** [comment]
Agreed. The minimum thing that would close this for me is a PR adding a napari-datashader example to the gallery.

**Comment by ahnsws** [comment]
Hello, I thought I'd revisit this. With some help from Claude Code, I was able to get 2D point clouds of sizes of up to 3e8 to work using napari, without using datashader. Here is a video of the marine AIS dataset (~212M points):

https://github.com/user-attachments/assets/ed49b58e-61e3-4241-915e-8e0e6793efa3

[Here is a standalone gist](https://gist.github.com/ahnsws/721d0b040802babb820f8dac34631ec9) which has 3e8 points.

The GPU I'm using is an NVIDIA GeForce RTX 3070 Ti (8192 MiB). The number of points able to be rendered seems to be determined by the amount of GPU memory, which makes total sense. Essentially, we had to use minimal shaders to be able to fit the dataset onto the GPU; if certain features are removed (e.g. borders, symbols, selecting/editing points), some performance can be gained; in the gist, changing face color, opacity, and marker size are supported. I suspect different colors per point can be done efficiently somehow using lookup tables, but I haven't investigated that further; I suppose I'd just add another layer with a different color if it were another class of points (for now). Edit: this would still require all points across classes to be on the GPU, so the point is moot hm.

I should note that there is some slight jitter in the points at high zoom in the video, which I don't see in the standalone gist.

I also tried rendering ~5e8 and ~1e9 points, but it resulted in a significantly laggier machine; maybe the dataset ended up being stored in CPU memory.

It's not buttery smooth (~100-500 ms per frame, depending on the size of the napari window), but it is much faster than the previous ~10 seconds per frame using datashader (I tried incorporating cupy/cudf, but it didn't really result in substantial performance gains).

**Comment by psobolewskiPhD** [comment]
Nice! thanks for the update!
@brisvag @aganders3 I wonder if any of that can be applied to napari/vispy? I think a programmatic only, read-only-ish (no selection/highlights) layer like this makes some sense. Apropos i'm fighting highlight performance in Shapes and now wondering the same thing there...

Edit Also super cool visualization. Took me a good long while to realize what it was showing, but once I got it.... 🤯

**Comment by aganders3** [comment]
Very interesting and really cool demo! We can look at optimizations in vispy but it feels more like a napari UX question if it's going to be a layer with reduced functionality. I'll definitely play with this though!

**Comment by Czaki** [comment]
I start thinking about introducing multiscale into points and shapes layers (or in general for all layers). This will allow us to limit the number of points sent to the GPU.

**Comment by aganders3** [comment]
Multi-scale feels like an optimization maybe better fit for napari without changing the UX.

**Comment by brisvag** [comment]
Yeah I think something like this screams points multiscale! In general it would be cool if napari could automagically create multiscale pyramids for single-scale data (which we should do for images as well), so that the user is not expected to generate it themselves and get still get the benefits of something like this.
